File: Material_RecyclerView/app/src/main/java/com/example/material_recyclerview/MainActivity.java
package com.example.material_recyclerview;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    private RecyclerView recyclerView;
    private RecyclerView.Adapter adapter;
    private RecyclerView.LayoutManager layoutManager;

    int[] list = {R.drawable.ic_baseline_emoji_emotions_24, R.drawable.ic_baseline_emoji_events_24,R.drawable.ic_outline_add_reaction_24};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ArrayList<RecyclerItem> recyclerItems = new ArrayList<>();
        recyclerItems.add(new RecyclerItem(R.drawable.ic_baseline_emoji_emotions_24, "Jonny", "friend"));
        recyclerItems.add(new RecyclerItem(R.drawable.ic_baseline_emoji_events_24, "Emmy", "girlfriend"));
        recyclerItems.add(new RecyclerItem(R.drawable.ic_outline_add_reaction_24, "Nick", "second friend"));

        for (int i=0; i<=2000; i++) {
            recyclerItems.add(new RecyclerItem(list[i%3], "name " + i, "relationship " + i));
        }

        recyclerView = findViewById(R.id.recyclerview);
        recyclerView.setHasFixedSize(true);
        adapter = new RecyclerAdapter(recyclerItems);
        layoutManager = new LinearLayoutManager(this);

        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(layoutManager);
    }
}

File: ContactsMVVM/app/src/main/java/com/example/contactsmvvm/MainActivity.kt
package com.example.contactsmvvm

import android.app.Dialog
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.Window
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import androidx.recyclerview.widget.RecyclerView.ViewHolder




class MainActivity : AppCompatActivity(), ContactRVAdapter.ContactItemClick,
    ContactRVAdapter.ContactDeleteItemClick {
    lateinit var itemsRV: RecyclerView
    lateinit var addFAB: FloatingActionButton
    private lateinit var list: List<Contact>
    private lateinit var contactRVAdapter: ContactRVAdapter
    lateinit var contactViewModel: ContactViewModel

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater: MenuInflater = menuInflater
        inflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.delete_all -> {
                contactViewModel.deleteAll()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        itemsRV = findViewById(R.id.contact_recycler_view)
        addFAB = findViewById(R.id.fab)
        list = ArrayList<Contact>()
        contactRVAdapter = ContactRVAdapter(list, this, this)
        itemsRV.layoutManager = LinearLayoutManager(this)
        itemsRV.adapter = contactRVAdapter
        val contactRepository = ContactRepositoryImplementation(ContactDatabase.invoke(this).dao)
        val contactFactory = ContactViewModelFactory(contactRepository)

        contactViewModel = ViewModelProvider(this, contactFactory)[ContactViewModel::class.java]

        contactViewModel.getAllContacts().observe(this, {
            contactRVAdapter.list = it
            contactRVAdapter.notifyDataSetChanged()
        })

        ItemTouchHelper(
            object : ItemTouchHelper.SimpleCallback(
                0,
                ItemTouchHelper.LEFT
            ) {
                override fun onMove(
                    recyclerView: RecyclerView,
                    viewHolder: ViewHolder, target: ViewHolder
                ): Boolean {
                    return false // true if moved, false otherwise
                }

                override fun onSwiped(viewHolder: ViewHolder, direction: Int) {
                   val contact =  contactRVAdapter.list[viewHolder.adapterPosition]
                    contactViewModel.deleteContact(contact)
                }
            }).attachToRecyclerView(itemsRV)

        addFAB.setOnClickListener {
            contactViewModel.openDialog(this)
        }
    }

    override fun onClickDeleteIcon(contact: Contact) {
        contactViewModel.deleteContact(contact)
        contactRVAdapter.notifyDataSetChanged()
        Toast.makeText(applicationContext, "Contacts has been delete", Toast.LENGTH_LONG).show()
    }

    override fun onClickContact(contact: Contact) {
        contactViewModel.openDialog(this, contact)
    }


}